### ISSUE TITLE
Immediately fail migration upon error in for await (.stream())

### ISSUE THREAD

**Issue by matthew-white** [getodk/central-backend#378]
In v1.2.0, one of the migrations runs `Form.fromXml()`, and if `Form.fromXml()` throws an error, the migration will run indefinitely rather than immediately failing. `Form.fromXml()` can throw an error if a form was uploaded before v1.1 with an ID ending in `.xml`/`.xls`/`.xlsx`.

https://github.com/getodk/central-backend/blob/24d20de20fdf2e03b55e687542fbc0cae889a9ff/lib/model/migrations/20210423-01-add-name-to-form-def.js#L24-L25

For future migrations, is there a way to get a migration to immediately fail if an error is thrown and to surface that error?

**Comment by issa-tseng** [comment]
there is no way to guarantee this as long as you are doing things w promises and timers. maybe you can add a total migration timeout but maybe it's just running a long time bc the db is big.

**Comment by matthew-white** [comment]
I took another look at this, and I think I have more information.



> In v1.2.0, one of the migrations runs `Form.fromXml()`, and if `Form.fromXml()` throws an error, the migration will run indefinitely rather than immediately failing.

Just to add more context to this, I think I remember that the migration would run indefinitely both in production and locally.

I think part of the issue here may be our use of `for await (.stream())`. For example, if I write a migration like this:

```js
const up = (db) => {
  throw new Error('some error');
};

const down = (db) => {};

module.exports = { up, down };
```

Then when I run migrations, I immediately get an error message:

```
node -e 'const { withDatabase, migrate } = require("./lib/model/migrate"); withDatabase(require("config").get("default.database"))(migrate);'
migration file "20210909-01-foo.js" failed
migration failed with error: some error
(node:89429) UnhandledPromiseRejectionWarning: Error: some error
    at Object.up (/central-backend/lib/model/migrations/20210909-01-foo.js:2:9)
    at /central-backend/node_modules/knex/lib/migrate/Migrator.js:529:40
    at processTicksAndRejections (internal/process/task_queues.js:85:5)
(node:89429) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:89429) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

I see the same results when the migration returns a promise:

```js
const up = (db) => Promise.reject(new Error('some error'));
```

However, I get a different result if I add the `for await` from `20210423-01-add-name-to-form-def.js`:

```js
const up = async (db) => {
  for await (const def of db.select('*').from('form_defs').stream()) {
    throw new Error('some error');
  }
};
```

Instead of immediately displaying an error message, the migration runs indefinitely. Is there a different way to stream that would fail immediately? Maybe we should consider using a try/catch within any future `for await (.stream())`?

That's also maybe a question we could look at the next time we write a migration like this. Since this issue seems to come up only for specific migrations, there's no change that's needed today, so it might make sense to close this issue for now.

**Comment by issa-tseng** [comment]
try/catch and for await don’t go well together. we’d probably need to rewrite to not use for await. can we change the ticket of the title to be specific to this migration? i don’t think there are general solutions possible here, as you see from how this case failed.

**Comment by matthew-white** [comment]
The specific migration was fixed as part of #379, so that migration shouldn't ever result in an error. The migration still uses `for await (.stream())`, but nothing in that for-loop should throw an error. But I've updated the title of this issue to reflect that the question now isn't about migrations in general, just those that use `for await (.stream())`. We've had three of those so far:

https://github.com/getodk/central-backend/blob/f611100f649d67c20110364bd206d5fd099af4d7/lib/model/migrations/20200220-01-repair-submission-parsing.js#L13-L29
https://github.com/getodk/central-backend/blob/f611100f649d67c20110364bd206d5fd099af4d7/lib/model/migrations/20210120-01-instance-names.js#L17-L26
https://github.com/getodk/central-backend/blob/f611100f649d67c20110364bd206d5fd099af4d7/lib/model/migrations/20210423-01-add-name-to-form-def.js#L23-L31

**Comment by matthew-white** [comment]
> However, I get a different result if I add the `for await` from `20210423-01-add-name-to-form-def.js`. &hellip; Instead of immediately displaying an error message, the migration runs indefinitely.

I'm actually not able to reproduce this now, either locally or on a QA server. Even if I throw an error from within a `for await`, the migration fails immediately and displays the error message. I don't know why things are different now (is it our move to Node 14?), but this is very good news for troubleshooting migrations. There are still some mysteries around `for await` and migrations (see #429), but at least this mystery seems to be resolved.
